File: 98.validate-binary-search-tree.java
/*
 * @lc app=leetcode id=98 lang=java
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean res = true;
    private long prev = Long.MIN_VALUE;
    public boolean isValidBST(TreeNode root) {
        valid(root);
        return res;
    }

    public void valid(TreeNode node) {
        if (node == null) {
            return;
        }

        valid(node.left);
        if (node.val <= prev) {
            res = false;
        }
        prev = node.val;
        valid(node.right);
    }
}
// @lc code=end

File: code/98.validate-binary-search-tree.go
/*
 * @lc app=leetcode id=98 lang=golang
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func check(root *TreeNode, max *int) bool {
	if root == nil {
		return true
	}

	left := check(root.Left, max)
	if !left {
		return false
	}

	if *max >= root.Val {
		return false
	}
	*max = root.Val

	right := check(root.Right, max)
	return right
}

func isValidBST(root *TreeNode) bool {
	prev := math.MinInt64
	return check(root, &prev)
}

// @lc code=end

